lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp

```diff
@@ -0,0 +1,307 @@
+    callee = tryCallOp.getCalleeAttr().getAttr();
+    if (callee != StringAttr::get(op->getContext(), DNNL_BRGEMM_DISPATCH_NAME))
+      return nullptr;
+    return tryCallOp;
+  } else if (auto tryLoadOp = dyn_cast_or_null<LLVM::LoadOp>(kernelProducer)) {
+    auto tryAddrOfOp = dyn_cast_or_null<LLVM::AddressOfOp>(
+        tryLoadOp.getOperand().getDefiningOp());
+    if (!tryAddrOfOp)
+      return nullptr;
+    return traceDispatchInGlobalCtor(module, tryAddrOfOp.getGlobalName());
+  }
+  return nullptr;
+}
+
+Operation *
+BrgemmDispatchAnalysis::traceDispatchInGlobalCtor(ModuleOp module,
+                                                  llvm::StringRef global_name) {
+  std::string gctor_name = std::string(global_name) + "_ctor";
+  FlatSymbolRefAttr ctorName =
+      SymbolRefAttr::get(module->getContext(), gctor_name);
+  auto ctor = module.lookupSymbol<LLVM::LLVMFuncOp>(ctorName);
+  if (!ctor)
+    return nullptr;
+
+  // ctor should contain only one call for kernel dispatch
+  auto &body = ctor.getBody();
+  for (auto &opRef : body.getOps()) {
+    auto *op = &opRef;
+    auto tryCallOp = dyn_cast_or_null<func::CallOp>(op);
+    if (!tryCallOp)
+      continue;
+    auto callee = tryCallOp.getCalleeAttr().getAttr();
+    if (callee == StringAttr::get(op->getContext(), DNNL_BRGEMM_DISPATCH_NAME))
+      return op;
+  }
+  return nullptr;
+}
+
+// return a pair of <tilecfg Op, tilerelease Op> extracted from given region,
+// only check direct descendants
+static std::pair<Operation *, Operation *>
+extractTileOpsFromRegion(Region &region) {
+  std::pair<Operation *, Operation *> ret{nullptr, nullptr};
+
+  for (auto &opRef : region.getOps()) {
+    LLVM_DEBUG(llvm::dbgs() << ">>> " << opRef << "\n");
+    auto *op = &opRef;
+    auto tryCallOp = dyn_cast_or_null<func::CallOp>(op);
+    if (!tryCallOp)
+      continue;
+    auto callee = tryCallOp.getCalleeAttr().getAttr();
+    if (callee == StringAttr::get(op->getContext(), DNNL_BRGEMM_TILECFG_NAME))
+      ret.first = op;
+    else if (callee ==
+             StringAttr::get(op->getContext(), DNNL_BRGEMM_TILERELEASE_NAME))
+      ret.second = op;
+  }
+
+  return ret;
+}
+
+static bool dispatchHasSameContext(Operation *lhs, Operation *rhs) {
+  auto lhsDispatch = dyn_cast_or_null<func::CallOp>(lhs);
+  auto rhsDispatch = dyn_cast_or_null<func::CallOp>(rhs);
+  if (!lhsDispatch || !rhsDispatch)
+    return false;
+  auto dispatchNameAttr =
+      StringAttr::get(lhs->getContext(), DNNL_BRGEMM_DISPATCH_NAME);
+  if (lhsDispatch.getCalleeAttr().getAttr() != dispatchNameAttr ||
+      rhsDispatch.getCalleeAttr().getAttr() != dispatchNameAttr)
+    return false;
+
+  auto lhsOperands = lhsDispatch.getOperands();
+  auto rhsOperands = rhsDispatch.getOperands();
+  assert(lhsOperands.size() == rhsOperands.size() &&
+         "Inconsistent operand size");
+  for (size_t idx = 0; idx < lhsOperands.size(); idx++) {
+    if (idx == 8) {
```

Comment:
Please use a meaningful constant variable name to replace the `8`

Reply:
fixed.